File: CSharpCodeBase/components/controlsystem/action.cs
namespace MainGame
{
    public class action
    {
        protected int priority;
        protected bool continuous;
        protected string name;

        private bool isStarted = false;

        public void init(int priority, bool continuous, string name)
        {
            this.priority = priority;
            this.continuous = continuous;
            this.name = name;
            this.isStarted = false;
        }

        public bool IsStarted()
        {
            return isStarted;
        }

        public void OnStart()
        {
            isStarted = true;
            OnStartRunning();
        }

        public void OnSuspend()
        {
            OnStopRunning();
        }

        public void OnResume()
        {
            OnStartRunning();
        }

        public void OnComplete()
        {
            OnStopRunning();
        }

        public void OnRemove()
        {
            OnStopRunning();
        }

        virtual public void FixedUpdate()
        {
        }

        virtual public bool Update()
        {
        }

        virtual public void OnStopRunning()
        {
        }

        virtual public void OnStartRunning()
        {
        }

        virtual public void OnPushed()
        {
        }

        virtual public bool IsContinuous()
        {
            return continuous;
        }

        virtual public int GetPriority()
        {
            return priority;
        }

        virtual public void OnEvent(object data)
        {
        }
    }
}
 

File: CSharpCodeBase/components/controlsystem/action.lua
local Class = require 'utils.hump.class' 

local Action = Class({init=function(self, priority, continuous, name)
      self.priority = priority
      self.continuous = continuous and true or false
      self.isStarted = false
      self.name = name
end})

function Action:OnFetch(...)
  self:init(...)
end

function Action:Free()
  if self.__pool ~= nil then
    self.__pool:Release(self)
  end
end

function Action:IsStarted()
  return self.isStarted
end

function Action:OnStart()
  self.isStarted = true
  self:OnStartRunning()  
end

function Action:OnSuspend()
  self:OnStopRunning()
end

function Action:OnResume()
  self:OnStartRunning()
end

function Action:OnComplete()
  self:OnStopRunning()
end

function Action:FixedUpdate()
  
end

function Action:OnStopRunning()
end

function Action:OnStartRunning()
end

function Action:OnPushed()
end

function Action:OnRemove()
  self:OnStopRunning()
end

function Action:IsContinuous()
  return self.continuous
end

function Action:GetPriority()
  return self.priority 
end

function Action:OnEvent(data)
end

return Action